### ISSUE TITLE
Must use rc4-hmac encryption?

### ISSUE THREAD

**Issue by jallakim** [modauthgssapi/mod_auth_gssapi#145]
Hi,

Trying to get SSO ("Windows Integrated Authentication") to work on a site hosted on Apache (Linux, RHEL 7.3). Tried first with mod_auth_kerb, but found that mod_auth_gssapi might be better suited (more updated, etc).

I've done all the "needful" (setspn, ktpass, etc), and have the keytab-file;

```
[root@somehost sites-available]# klist -k somesite.foo.bar.keytab -e
Keytab name: FILE:somesite.foo.bar.keytab
KVNO Principal
---- --------------------------------------------------------------------------
   4 HTTP/somesite.foo.bar@FOO.BAR (aes256-cts-hmac-sha1-96) 

[root@somehost sites-available]# kinit -k -t somesite.foo.bar.keytab HTTP/somesite.foo.bar@FOO.BAR

[root@somehost sites-available]# klist 
Ticket cache: KEYRING:persistent:0:0
Default principal: HTTP/somesite.foo.bar@FOO.BAR

Valid starting     Expires            Service principal
14/07/17 17:34:39  15/07/17 03:34:39  krbtgt/FOO.BAR@FOO.BAR
	renew until 21/07/17 17:34:39
```

Then the following configuration in the vhost;

```
# AD SSO
AuthType GSSAPI
AuthName "somesite.foo.bar AD SSO"
GssapiCredStore keytab:/etc/httpd/sites-available/somesite.foo.bar.keytab 
require valid-user
```

Yielding the following errors in the log;

```
[Fri Jul 14 17:38:09.421898 2017] [auth_gssapi:error] [pid 38434] [client <snip>] NO AUTH DATA Client did not send any authentication headers
[Fri Jul 14 17:38:09.446039 2017] [auth_gssapi:error] [pid 38434] [client <snip>] GSS ERROR In Negotiate Auth: gss_accept_sec_context() failed: [Unspecified GSS failure.  Minor code may provide more information (Request ticket server HTTP/somesite.foo.bar@FOO.BAR kvno 4 found in keytab but not with enctype rc4-hmac)]
```

I was under the impression that ```-crypto RC4-HMAC-NT``` (as the ```ktpass.exe``` parameter) only was needed when/if not all AD servers where 2008 or newer? In our case, we only have 2008 and newer, and hence I used ```-crypto AES256-SHA1``` as the parameter. But judging by the logs, it seems as we *must* use ```rc4-hmac``` as the encryption type?

I am not that "into" Kerberos and all these things, so sorry if this is a stupid question and/or not mod_auth_gssapi related at all.

**Comment by frozencemetery** [comment]
Well... rc4 is no longer considered secure, so if there's a way to avoid using that altogether, it's recommended.  @simo5, do you know what the status of AD and ciphers is these days?

**Comment by simo5** [comment]
Ad supports AES, and should use it by default IF you have set a domain level > 2008 (domain level, not server version).
That said clients may still fall back to rc4 if they are old enough I guess.

**Comment by frenche** [comment]
You'd have to enable the corresponding aes check box on the service account in AD, otherwise the KDC won't assume the service has aes keys when a user asks for a tgs.

**Comment by jallakim** [comment]
> Well... rc4 is no longer considered secure, so if there's a way to avoid using that altogether, it's recommended.

Yes, that was why I tried to go the AES route.

So, I need to both have ```domain level > 2008``` *and* enable AES on the user, and even then, you can end up having (old) clients only supporting rc4? Sigh. I guess "rc4 it is", then :-|

**Comment by jallakim** [comment]
So, the domain level for this domain is "2008 R2". Not sure why it complains about no rc4 then? The client is a Mac OS X (Sierra, 10.12.3) running Heimdal 1.5.1apple1. The user ticket on the client is encryption ```aes256-cts-hmac-sha1-96```, and has successfully gotten ticket on other SSO sites using that encryption level.

In can also see entries on the client for the SSO site I'm trying to set up;

```
Server: HTTP/somesite.foo.bar@FOO.BAR
Client: user@FOO.BAR
Ticket etype: arcfour-hmac-md5, kvno 4
Ticket length: 2728
Auth time:  Jul 17 12:30:18 2017
Start time: Jul 17 12:30:24 2017
End time:   Jul 17 22:30:15 2017
Ticket flags: pre-authent, forwardable
Addresses: addressless

Server: HTTP/somesite.foo.bar@FOO.BAR
Client: user@FOO.BAR
Ticket etype: arcfour-hmac-md5, kvno 2
Ticket length: 2503
Auth time:  Jul 17 12:30:18 2017
Start time: Jul 17 12:31:22 2017
End time:   Jul 17 22:30:15 2017
Ticket flags: enc-pa-rep, pre-authent, forwardable
Addresses: addressless
```

Not sure why the client tries to use ```arcfour-hmac-md5``` when it seemingly should be able to use AES?

**Comment by frenche** [comment]
Again, you have to enable the corresponding AES check-box on the service account (the principal representing HTTP/somesite.foo.bar service).

**Comment by jallakim** [comment]
> Again, you have to enable the corresponding AES check-box on the service account (the principal representing HTTP/somesite.foo.bar service).

Ah, sigh, I completely forgot that part. I'll try that now.

**Comment by jallakim** [comment]
That did the trick. Sorry again, and thanks (-:

**Comment by jallakim** [comment]
For future reference; I also got the same error as in #13;

```
GSS ERROR In Negotiate Auth: gss_accept_sec_context() failed: [An unsupported mechanism was requested (Unknown error)]
```

This only occurred on Windows-machines, even if they had valid Kerberos-tickets. They seemed to fall back to using NTLM, while on Linux/OS X it worked flawlessly. The culprit turned out to be that the FQDN used was a CNAME, and not a direct A/AAAA record. Not sure why Windows is so "picky" about it when it works fine in Linux/OS X, but changing it to an A/AAAA record solved the issue.

**Comment by simo5** [comment]
Windows is correct, we tried for a long time to change the default of dns_canonicalize_hostname to false,
you can read here why trusting DNS to canonicalize or resolve PTR records is quite a bad idea:
https://ssimo.org/blog/id_015.html
Note that CNAME -> A has exactly the same issues as PTR -> A

HTH.

**Comment by jallakim** [comment]
Why does CNAME --> A have the same issue? At least when both both the CNAME- and A-record is the same domain/TLD (which is usually the case in corporate environments), you'd have the same opportunity to "intercept"/abuse regardless if it's CNAME or A-records? The problem with PTRs is the two-way lookup (take the FQDN, resolve IP, then do PTR on that, and use the new hostname provided).

Sure, you could have the same issue with CNAME's if they are cross-domain, but that could be solved programmatically by only allowing CNAMEs if they are the same TLD (or use some kind of whitelist provided in the Kerberos-config or whatnot).

**Comment by simo5** [comment]
client:
- DNS please tell me what does secret.server.my.domain point to ?
attacker (MITM of DNS reply):
- it is a CNAME and points to points to A[attacker.controlled.server.my.domain]
client:
- oh gee thanks getting ticket for attacker.controlled.server.my.domain and connecting to it

The only saving grace with mod_auth_gssapi is that you are *supposed* to use HTTPS therefore hopefully your client will not just happily also resolve the CNAME and connect to a random server but will just use the fqdn as passed in and verify the server cert matches that name.
So the client will connect to the "right" server with the wrong ticket and just fail.

OTOH in many copr networks ppl are used to just click ok on any TLS cert error, or all machines are given a wildcard cert so back to being potentially vulnerable ...

**Comment by jallakim** [comment]
But is that in reality any different than;

1. DNS please tell me what does secret.server.my.domain point to?
2. Attacker (MITM) replies: it's an A record that resolves to 10.10.10.10 (attacker controller host, where 10.20.20.0 would be the valid reply).

If you can MITM the client DNS response, there is surely not much difference between CNAME and A-records, unless I've missed something stupid here?

**Comment by simo5** [comment]
Ys it is very different from just redirecting to a random IP address, because when you do that all that you obtain is an authentication failure (if the client is performing mutual authentication of course).
As the client will obtain a ticket for secret.server which the attacker.controlled host can't deal with.

While in the case where you MITM CNAME (or PTR) resolution, as the client does canonicalaization, then it is going to get a ticket for attacker.controlled host (without the user noticing unless they examine the cred cache). Note that this is a legitimate host on the network that has been taken over, not a random machine on the internet with no credentials in the realm.

DNSSEC is a good solution to this problem, as it cannot be MITMd. In fact it would be nice to have a way to do canonicalization only conditional to DNSSEC protection, but we do not have interfaces for that yet, and most corporate networks don't do DNSSEC anyway.
